SandWorm/MeshCompareComponent.cs

```diff
@@ -0,0 +1,174 @@
+            pManager.AddGenericParameter("SandWormOptions", "SWO", "Setup & Calibration options", GH_ParamAccess.item);
+            pManager[0].Optional = true;
+            pManager[1].Optional = false;
+            pManager[2].Optional = true;
+            pManager[3].Optional = true;
+        }
+
+        /// <summary>
+        /// Registers all the output parameters for this component.
+        /// </summary>
+        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
+        {
+            pManager.AddGeometryParameter("Surface", "S", "Sandbox surface in real-world scale", GH_ParamAccess.list);
+            pManager.AddPointParameter("Points", "P", "Additional mesh analysis", GH_ParamAccess.list);
+            pManager.AddGenericParameter("MeshCompareOptions", "MCO", "Resulting array of elevation points", GH_ParamAccess.item);
+        }
+
+        /// <summary>
+        /// This is the method that actually does the work.
+        /// </summary>
+        /// <param name="DA">The DA object is used to retrieve from inputs and store in outputs.</param>
+        protected override void SolveInstance(IGH_DataAccess DA)
+        {
+            options = new SetupOptions();
+            results = new CompareMeshes();
+
+            DA.GetData<Curve>(0, ref inputRectangle);
+            DA.GetData<Mesh>(1, ref inputMesh);
+            DA.GetData<double>(2, ref scaleFactor);
+            DA.GetData<SetupOptions>(3, ref options);
+
+            if (options.SensorElevation != 0) sensorElevation = options.SensorElevation; 
+            if (scaleFactor <= 0)
+                AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Scale Factor must be greater than 0.");
+            if (options.LeftColumns != 0) leftColumns = options.LeftColumns;
+            if (options.RightColumns != 0) rightColumns = options.RightColumns;
+            if (options.TopRows != 0) topRows = options.TopRows;
+            if (options.BottomRows != 0) bottomRows = options.BottomRows;
+
+
+            // Shared variables
+            var unitsMultiplier = Core.ConvertDrawingUnits(Rhino.RhinoDoc.ActiveDoc.ModelUnitSystem);
+            sensorElevation /= unitsMultiplier; // Standardise to mm to match sensor units
+            Core.PixelSize depthPixelSize = Core.GetDepthPixelSpacing(sensorElevation);
+            var trimmedWidth = (512 - leftColumns - rightColumns) * depthPixelSize.x * unitsMultiplier * scaleFactor;
+            var trimmedHeight = (424 - topRows - bottomRows) * depthPixelSize.y * unitsMultiplier * scaleFactor;
+
+            // Initialize all the outputs
+            output = new List<string>();
+            outputSurface = new List<GeometryBase>();
+            results.MeshElevationPoints = new double[(512 - leftColumns - rightColumns) * (424 - topRows - bottomRows)];
+            List<Mesh> inputMeshes = new List<Mesh>{ inputMesh };
```

Comment:
Why is this variable a list here? I assume its so that multiple meshes can be used to create the comparison terrain (as the projection method below handles) but if so it seems like the input type should be changed (and a note about this added in its help string?)

Reply:
The only reason for this being a list is because the `ProjectPointsToMeshes` method takes a list as an input.
https://developer.rhino3d.com/api/RhinoCommon/html/M_Rhino_Geometry_Intersect_Intersection_ProjectPointsToMeshes.htm

https://github.com/mariuszhermansdorfer/SandWorm/blob/93fb86ad80beb513f00d2bbaaaf98e0e0765b27e/SandWorm/MeshCompareComponent.cs#L128